neuralop/layers/codano.py

```diff
@@ -0,0 +1,462 @@
+        # We have an option to make the last operator (MLP in regular
+        # Transformer block) permutation equivariant. i.e., applying the
+        # operator per variable or applying the operator on the whole channel
+        # (like regular FNO).
+        if permutation_eq:
+            self.mixer = FNOBlocks(
+                in_channels=self.mixer_token_codimension,
+                out_channels=self.mixer_token_codimension,
+                apply_skip=True,
+                n_layers=2,
+                **mixer_args,
+                **common_args,
+            )
+            self.norm1 = Normalizer(self.token_codimension)
+            self.norm2 = Normalizer(self.mixer_token_codimension)
+            self.mixer_out_normalizer = Normalizer(
+                self.mixer_token_codimension)
+
+        else:
+            self.mixer = FNOBlocks(
+                in_channels=codimension_size,
+                out_channels=codimension_size,
+                n_layers=2,
+                **mixer_args,
+                **common_args,
+            )
+            self.norm1 = Normalizer(codimension_size)
+            self.norm2 = Normalizer(codimension_size)
+            self.mixer_out_normalizer = Normalizer(codimension_size)
+
+    def forward(self, *args):
+        raise NotImplementedError(
+            "Use a proper subclass of TNOBlock (i.e. TnoBlock2d or TNOBlock3D).")
+
+
+class CODABlocks2D(CODABlocks):
+    def __init__(self, *args, **kwargs):
+        Normalizer = kwargs.get("Normalizer")
+        if Normalizer is None:
+            Normalizer = AffineNormalizer2D
+        kwargs["Normalizer"] = Normalizer
+
+        Convolution = kwargs.get("SpectralConvolution")
+        if Convolution is None:
+            Convolution = SpectralConv2d
+        kwargs["SpectralConvolution"] = Convolution
+
+        super().__init__(*args, **kwargs)
+
+    # XXX rewrite comments on TNO*3D
+    def compute_attention(self, xa, batch_size):
+        """Compute the key-query-value variant of the attention matrix.
+
+        Assumes input ``xa`` has been normalized.
+        """
+        k = self.K(xa)
+        q = self.Q(xa)
+        v = self.V(xa)
+
+        value_x, value_y = v.shape[-2], v.shape[-1]
+
+        rearrangement = dict(
+            pattern='(b t) (a d) h w -> b a t (d h w)',
+            b=batch_size,
+            a=self.n_head,
+        )
+        k = rearrange(k, **rearrangement)
+        q = rearrange(q, **rearrangement)
+        v = rearrange(v, **rearrangement)
+
+        dprod = (torch.matmul(q, k.transpose(-1, -2)) /
+                 (np.sqrt(k.shape[-1]) * self.temperature))
+        dprod = F.softmax(dprod, dim=-1)
+
+        attention = torch.matmul(dprod, v)
+        attention = rearrange(
+            attention,
+            'b a t (d h w) -> b t a d h w',
+            d=self.head_codimension,
+            h=value_x,
+            w=value_y,
+        )
+        attention = rearrange(attention, 'b t a d h w -> (b t) (a d) h w')
```

Comment:
Why are we doing two rearrange one after the other on the same variable?

Reply:
Thanks for noticing. It can be by one rearrange operation.